src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +36,455 @@ static int16_t axisval(int16_t x)
-
-        joystate[joy].instance = instance;
+    if (got_event) {
+        last_poll_ = now;
+    } else if (now - last_poll_ > POLL_TIME_MS) {
+        do_poll = true;
+        last_poll_ = now;
     }
 
-    joystate[joy].index = joy;
-    joystate[joy].is_gc = is_gc;
-}
-
-void wxSDLJoy::RemapControllers()
-{
-    for (auto&& joy : joystate) {
-        auto& state = joy.second;
-
-        DisconnectController(state);
-        ConnectController(joy.first);
+    if (do_poll) {
+        for (auto&& joy_state : joystick_states_) {
+            joy_state.second->Poll();
+        }
     }
 }
 
-void wxSDLJoy::DisconnectController(wxSDLJoyState& state)
-{
-    if (auto& dev = state.dev) {
-        if (state.is_gc) {
-            SDL_GameControllerClose(dev);
-        }
-        else {
-            SDL_JoystickClose(dev);
-        }
-
-        dev = nullptr;
+void wxSDLJoy::ConnectController(int index) {
+    std::unique_ptr<wxSDLJoyState> joy_state(new wxSDLJoyState(index));
+    if (joy_state->IsValid()) {
+        joystick_states_.emplace(
+            joy_state->joystick_id(), std::move(joy_state));
     }
-
-    instance_map.erase(state.instance);
 }
 
-void wxSDLJoy::Add(int8_t joy_n)
-{
-    if (joy_n < 0) {
-        for (uint8_t joy : range(0, SDL_NumJoysticks()))
-            ConnectController(joy);
-
-        add_all = true;
-
+void wxSDLJoy::PollJoystick(int index) {
+    if (add_all_joysticks_)
         return;
-    }
-
-    ConnectController(joy_n);
+    if (requested_indexes_.insert(index).second)
+        ConnectController(index);
 }
 
-void wxSDLJoy::Remove(int8_t joy_n)
-{
-    add_all = false;
+void wxSDLJoy::PollAllJoysticks() {
+    requested_indexes_.clear();
+    add_all_joysticks_ = true;
+    joystick_states_.clear();
 
-    if (joy_n < 0) {
-        for (auto&& joy : joystate)
-            DisconnectController(joy.second);
-
-        joystate.clear();
-
-        return;
-    }
+    for (int i = 0; i < SDL_NumJoysticks(); i++)
+        ConnectController(i);
+}
 
-    DisconnectController(joystate[joy_n]);
-    joystate.erase(joy_n);
+void wxSDLJoy::StopPolling() {
+    add_all_joysticks_ = false;
+    requested_indexes_.clear();
+    joystick_states_.clear();
 }
 
-void wxSDLJoy::SetRumble(bool do_rumble)
-{
-    rumbling = do_rumble;
+void wxSDLJoy::SetRumble(bool activate_rumble) {
+    rumbling_ = activate_rumble;
+    if (joystick_states_.empty())
+        return;
 
 #if SDL_VERSION_ATLEAST(2, 0, 9)
-    // Do rumble only on device 0, and only if it's a GameController.
-    auto dev = joystate[0].dev;
-    if (dev && SDL_IsGameController(0)) {
-        if (rumbling) {
-            SDL_GameControllerRumble(dev, 0xFFFF, 0xFFFF, 300);
-            if (!IsRunning())
-                Start(150);
-        }
-        else {
-            SDL_GameControllerRumble(dev, 0, 0, 0);
+    // Do rumble only on the first device.
+    joystick_states_.begin()->second->SetRumble(activate_rumble);
```

Comment:
Please do the check for the first device being a GC here, we don't want to run the timer unnecessarily.

Reply:
I changed this so it is handled in the `wxSDLJoyState` class. There is no need to do this in `wxSDLJoy` class anymore.